File: backend/app/api/v1/resources.py
"""资源站 API"""
from fastapi import APIRouter

from app.schemas import resources as schemas_resources
from app.schemas.response import ApiResponse, error, success
from app.services import resources as services_resources
from app.utils.exception import ServiceException


router = APIRouter(prefix="/resources", tags=["resources"])


@router.post(
    "/sites/list",
    response_model=ApiResponse[schemas_resources.ResourceSiteListResponse],
)
def query_resource_site_list(_request: schemas_resources.ResourceSiteListRequest):
    """查询资源站列表"""
    try:
        result = services_resources.get_site_list()
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/sites/detail",
    response_model=ApiResponse[schemas_resources.ResourceSiteResponse],
)
def query_resource_site_detail(data: schemas_resources.ResourceSiteDetailRequest):
    """查询资源站详情"""
    try:
        result = services_resources.get_site_detail(data.site_id)
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/sites/toggle",
    response_model=ApiResponse[schemas_resources.ResourceSiteToggleResponse],
)
def toggle_resource_site(data: schemas_resources.ResourceSiteToggleRequest):
    """切换资源站启用状态"""
    try:
        result = services_resources.toggle_site_enabled(data.site_id)
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/sites/test",
    response_model=ApiResponse[schemas_resources.ResourceSiteTestResponse],
)
async def test_resource_site(data: schemas_resources.ResourceSiteTestRequest):
    """测试资源站连接"""
    try:
        result = await services_resources.test_site_connection(data.site_id)
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/sites",
    response_model=ApiResponse[schemas_resources.ResourceSiteResponse],
)
def create_resource_site(data: schemas_resources.ResourceSiteCreateRequest):
    """新增资源站"""
    try:
        result = services_resources.create_site(data)
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/sites/update",
    response_model=ApiResponse[schemas_resources.ResourceSiteResponse],
)
def update_resource_site(data: schemas_resources.ResourceSiteUpdateActionRequest):
    """更新资源站配置"""
    try:
        update_data = schemas_resources.ResourceSiteUpdateRequest.model_validate(
            data.model_dump(exclude={"site_id"}, exclude_unset=True)
        )
        result = services_resources.update_site(data.site_id, update_data)
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/sites/delete",
    response_model=ApiResponse[schemas_resources.ResourceSiteDeleteResponse],
)
def delete_resource_site(data: schemas_resources.ResourceSiteDeleteRequest):
    """删除资源站"""
    try:
        result = services_resources.delete_site(data.site_id)
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/sites/test-all",
    response_model=ApiResponse[schemas_resources.ResourceSiteBatchTestResponse],
)
async def test_all_resource_sites(
    _request: schemas_resources.ResourceSiteBatchTestRequest,
):
    """批量测试所有已启用资源站"""
    try:
        result = await services_resources.test_enabled_sites()
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/config/export",
    response_model=ApiResponse[schemas_resources.ResourceConfigExportResponse],
)
def export_resource_config(_request: schemas_resources.ResourceConfigExportRequest):
    """导出资源站配置"""
    try:
        result = services_resources.export_config()
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)


@router.post(
    "/config/import",
    response_model=ApiResponse[schemas_resources.ResourceConfigImportResponse],
)
def import_resource_config(data: schemas_resources.ResourceConfigImportRequest):
    """导入资源站配置"""
    try:
        result = services_resources.import_config(data)
        return success(data=result)
    except ServiceException as exc:
        return error(code=exc.code, message=exc.message)

File: frontend/src/api/resources.js
import request from '@/utils/request'

export function getResourceSites() {
  return request.get('/api/v1/resources/sites')
}

export function getResourceSiteDetail(siteId) {
  return request.get(`/api/v1/resources/sites/${siteId}`)
}

export function toggleResourceSite(siteId) {
  return request.post(`/api/v1/resources/sites/${siteId}/toggle`)
}

export function testResourceSite(siteId) {
  return request.post(`/api/v1/resources/sites/${siteId}/test`)
}

/**
 * 创建资源站
 * @param {Object} siteData - 站点数据（snake_case：site_id/name/base_url/enabled/timeout/search_endpoint/page_param/action_param）
 * @returns {Promise} 新建站点
 */
export function createResourceSite(siteData) {
  return request.post('/api/v1/resources/sites', siteData)
}

/**
 * 更新资源站
 * @param {string} siteId - 站点 ID
 * @param {Object} siteData - 待更新字段的可选子集（snake_case，site_id 不可改）
 * @returns {Promise} 更新后站点
 */
export function updateResourceSite(siteId, siteData) {
  return request.put(`/api/v1/resources/sites/${siteId}`, siteData)
}

/**
 * 删除资源站
 * @param {string} siteId - 站点 ID
 * @returns {Promise} 删除结果
 */
export function deleteResourceSite(siteId) {
  return request.delete(`/api/v1/resources/sites/${siteId}`)
}

/**
 * 批量测试全部已启用资源站
 * @returns {Promise} 批量测试结果（total/success_count/failed_count/results）
 */
export function testAllResourceSites() {
  return request.post('/api/v1/resources/sites/test-all')
}

/**
 * 导出资源站配置
 * @returns {Promise} 配置数据（sites 数组）
 */
export function exportResourceConfig() {
  return request.get('/api/v1/resources/config/export')
}

/**
 * 导入资源站配置
 * @param {Array} sites - 站点配置数组（snake_case 字段）
 * @returns {Promise} 导入结果（imported_count/message）
 */
export function importResourceConfig(sites) {
  return request.post('/api/v1/resources/config/import', { sites })
}

